### PR TITLE
feat: api overhaul

### PR DESCRIPTION
**Pull request by alittlebitweird**
## Naming consistency changes
* `xyzAttribute` -> `xyzKey`
* `updateXYZ` -> `setXYZ`

### Layer Structure
* We now distinguish between `layers` and `layerConfigs`.

#### Layers
* audio
* filter
* html
* image
* lottie
* sequence
* subtitles
* text
* video
* waveform

#### Layer Configs
* audio
* background
* size
* trim
* timeline
* position

An `XYZLayer` is an object that is comprised of a combination of layer-specific `options` and `layerConfig`. Each `layer` has a different set of composed `layerConfigs` (or none at all), but they all use the same name-spaced configuration, ex: 

```typescript
export type TextLayerConfig = Background & Position & Size & Timeline & Trim
export type TextNamespace = { [LayerKey.text]: TextOptions }
export type TextLayer = TextLayerConfig & TextNamespace

```

### Mixins
* goodbye `VisualMedia` and any multi-level-deep inheritance for layers, all inheritance is done through `ts-mixer` now. This means we can be more modular about creating and extending layer classes, ex:
```typescript
export class Image extends Mixin(BackgroundMixin, PositionMixin, SizeMixin, TimelineMixin, TrimMixin) {

```

### Composition addXYZ methods
* all `addXYZ` methods adhere to the same general interface now: 
```typescript
  public async addXYZ(
    file: CompositionFile,
    options?: XYZOptions,
    layerConfig?: XYZLayerConfig
  ): Promise<XYZ> {
```
* some methods don't use all three `file`, `options`, or `layerConfig` arguments, but they are always in the same order, i.e.

```typescript
 public [CompositionMethod.addText](options: TextOptions, layerConfig?: TextLayerConfig): Text | undefined {


 public [CompositionMethod.addFilter]<FilterName extends keyof Filters>(options: {
   name: FilterName
   options?: Filters[FilterName]
 }): Filter | undefined {

```

```typescript
 public async [CompositionMethod.addVideo](file: CompositionFile, layerConfig?: VideoLayerConfig): Promise<Video> {
```

* Wherever possible, default arguments are set to reduce knowledge required by the user, via `_setLayerDefaults`, which also combines the `options` and `layerConfig` objects into the layer object that will ultimately be used to generate the final config:
```typescript
    const videoLayer: VideoLayer = this._setLayerDefaults<VideoLayer>({ layerConfig, layerType: LayerType.video })

```

### Validation
* vaidation is much more standardized, with the exception of `file` verification and `options` presence checking, all validation in `composition` is done via `layer` or `mixin`-specific validator functions that all call the generic`validateLayer`, ex:
```typescript
export const validateImageLayer = (layer: ImageLayer): void =>
  validateLayer<ImageLayer>(
    [validateBackground, validatePosition, validateSize, validateTimeline, validateTrim],
    CompositionMethod.addImage,
    layer
  )

export const validateTrimMixin = (callerName: TrimMethod, mixin: TrimOptions): void =>
  validateLayer<TrimOptions>([validateTrim], callerName, mixin)

```
* `options` in `addXYZ` functions are now runtime checked for correct types, and also validity of keys

### New stuff
* there are now `setXYZ` functions for every attribute on a `layer` or `mixin` and they all return `this` so they can be chained
* videos can set opacity / background color via `background` layer config

### CODE REVIEW

File: src/features/videos/layer/index.ts [hunk 1]
```diff
@@ -14,19 +13,20 @@ export class Layer {
     return this._id
   }
 
-  get start(): number | undefined {
-    return this._composition.getLayerAttribute<number | undefined>(this.id, LayerAttribute.start)
+  public getAttribute<Value>({ childKey, layerKey }: { childKey?: ChildKey; layerKey: LayerKey }): Value {
```
Comment (brycedev) [inline src/features/videos/layer/index.ts:16]: clean

File: src/features/videos/composition/index.ts [hunk 2]
```diff
@@ -131,42 +161,71 @@ export class Composition implements CompositionInterface {
...
+        return undefined
+      }
+    }
 
-  public [CompositionMethod.getLayerAttribute]<LayerAttributeValue>(
-    id: string,
-    layerAttribute: LayerAttribute
-  ): LayerAttributeValue {
-    return this._layers.find((layer) => layer.id === id)[layerAttribute]
+    return layer[layerKey]
```
Comment (ptrkcsk) [inline src/features/videos/composition/index.ts:185]: could we use optional chaining instead?:

```js
composition.layer(id)[LayerKey.text]?.[TextKey.text]
```

File: src/constant/validation/index.ts [hunk 3]
```diff
@@ -3,4 +3,4 @@ export interface ApiDataValidator<DataType> {
   validate: (data: unknown) => data is DataType
 }
 
-export type LayerValidator = (callerName: string, options: Record<string, any>) => string[]
+export type LayerValidator = (callerName: string, options: Record<string, any>, shouldThrow?: false) => string[]
```
[the remark this answers is not in the document]
Comment (alittlebitweird) [reply src/constant/validation/index.ts:6]: totally agree, i'll update this

File: src/constant/video/layers/html.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import { HTMLOptions } from 'constant/shared'
+
+export enum HTMLMethod {
```
[the remark this answers is not in the document]
Comment (alittlebitweird) [reply src/constant/video/layers/html.ts:3]: I'm cool with this. There are other examples like this, i.e. URL. I'll do a pass and update those as well

File: src/features/videos/composition/index.ts (hunk 2 above)
Comment (alittlebitweird) [reply src/features/videos/composition/index.ts:185]: excellent

Comment (alittlebitweird) [reply src/features/videos/composition/index.ts:185]: ^^^ this doesn't quite do what we need, but it pointed me in the right direction: 

```
    const layerAttribute = this._layers.find((layer) => layer.id === id)[layerKey]

    return childKey ? layerAttribute[childKey] : layerAttribute
```

**Pull request merged**